File: python/排序.py
def search(nums, target):
    left, right = 0, len(nums) - 1
    while left <= right:
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            right = mid - 1
        else:
            left = mid + 1

    return -1


def bubble_sort(nums):
    n = len(nums)
    for i in range(n - 1):
        flag = False
        for j in range(0, n - i - 1):
            if nums[j] > nums[j + 1]:
                nums[j], nums[j + 1] = nums[j + 1], nums[j]
                flag = True

        if not flag:
            break

    return nums


def select_sort(nums):
    n = len(nums)
    for i in range(n - 1):
        min_index = i
        for j in range(i + 1, n):
            if nums[j] < nums[min_index]:
                min_index = j

        nums[i], nums[min_index] = nums[min_index], nums[i]

    return nums


def insert_sort(nums):
    n = len(nums)
    for i in range(1, n):
        base = nums[i]
        j = i - 1
        while j >= 0 and nums[j] > base:
            nums[j + 1] = nums[j]
            j -= 1

        nums[j + 1] = base

    return nums


def quick_sort(nums):
    n = len(nums)
    if n < 2:
        return nums

    pivot = nums[0]
    left = [nums[i] for i in range(1, n) if nums[i] < pivot]
    right = [nums[i] for i in range(1, n) if nums[i] >= pivot]

    return quick_sort(left) + [pivot] + quick_sort(right)


if __name__ == "__main__":
    nums = [1, 3, 4, 5, 2, 7, 6]
    print(quick_sort(nums))

File: golang/排序.go
package main

import "fmt"

func search(nums []int, target int) int {
	n := len(nums)
	left, right := 0, n-1
	for left <= right {
		mid := (left + right) / 2
		if nums[mid] == target {
			return mid
		} else if nums[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return -1
}

func bubble_sort(nums []int) []int {
	n := len(nums)
	for i := 0; i < n-1; i++ {
		flag := false
		for j := 0; j < n-1-i; j++ {
			if nums[j] > nums[j+1] {
				nums[j], nums[j+1] = nums[j+1], nums[j]
				flag = true
			}
		}

		if !flag {
			break
		}
	}

	return nums
}

func select_sort(nums []int) []int {
	n := len(nums)
	for i := 0; i < n-1; i++ {
		min_index := i
		for j := i + 1; j < n; j++ {
			if nums[j] < nums[min_index] {
				min_index = j
			}
		}

		nums[i], nums[min_index] = nums[min_index], nums[i]
	}

	return nums
}

func insert_sort(nums []int) []int {
	n := len(nums)
	for i := 1; i < n; i++ {
		base := nums[i]
		j := i - 1
		for ; j >= 0 && nums[j] > base; j-- {
			nums[j+1] = nums[j]
		}

		nums[j+1] = base
	}

	return nums
}

func quick_sort(nums []int) []int {
	n := len(nums)
	if n < 2 {
		return nums
	}

	pivot := nums[0]
	left, right := make([]int, 0), make([]int, 0)
	for i := 1; i < n; i++ {
		if nums[i] < pivot {
			left = append(left, nums[i])
		} else {
			right = append(right, nums[i])
		}
	}

	return append(quick_sort(left), append([]int{pivot}, quick_sort(right)...)...)
}

func main() {
	nums := []int{1, 3, 5, 7, 9, 2, 4, 6, 8, 10}
	fmt.Println(search(nums, 10))
	fmt.Println(bubble_sort(nums))
	fmt.Println(select_sort(nums))
	fmt.Println(insert_sort(nums))
	fmt.Println(quick_sort(nums))
}
